### PR TITLE
[camera] Add support for Torch and Auto Exposure (#19845)

### PR DESCRIPTION
**Pull request by NeKoFu**
## Description

* Add new callable methods torchOn and torchOff
* Add new callable methods aeOn and aeOff
* Torch and AE mode could be activated at initialize

## Related Issues

Camera flash functionality during image and/or video capture [#19845](https://github.com/flutter/flutter/issues/19845)
Future of the Camera Plugin (Refactor/Rework) [#31225](https://github.com/flutter/flutter/issues/31225)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by NeKoFu** [comment]
To reviewers, sorry for the quantity of commits to review. I don't have a macbook to validate the iOS part and flutter_plugin_tools doesn't works on my windows. I used the CI to fix iOS code and code format. This feature works pretty well on android so I hope it will works also without any issues on iOS.

File: packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 1]
```diff
@@ -443,6 +460,42 @@ private void setImageStreamImageAvailableListener(final EventChannel.EventSink i
...
+    setTorchMode(result, enable, 1.0);
+  }
+
+  public void setTorchMode(@NonNull final Result result, boolean enable, double level) {
+    try {
+      captureRequestBuilder.set(
+          CaptureRequest.FLASH_MODE,
+          enable ? CaptureRequest.FLASH_MODE_TORCH : CaptureRequest.FLASH_MODE_OFF);
+
+      captureRequestBuilder.set(
```
Comment (mklim) [inline packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java]: Why are we always setting this now no matter the `FLASH_MODE`?

File: packages/camera/lib/camera.dart [hunk 2]
```diff
@@ -242,6 +244,12 @@ class CameraController extends ValueNotifier<CameraValue> {
   /// Whether to include audio when recording a video.
   final bool enableAudio;
 
+  // Torch
+  final bool enableTorch;
```
Comment (mklim) [inline packages/camera/lib/camera.dart]: nit: This and `enableAE` should explain more what these members symbolize, and should also be `/// ` to show that they're documentation and not implementation comments.

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#documentation-dartdocs-javadocs-etc

**Comment by malikdoksoz** [comment]
After taking a few pictures on Android, the picture does not appear and the camera cannot be used again.

File: packages/camera/CHANGELOG.md [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ## 0.5.3
-
 * Added new quality presets.
 * Now all quality presets can be used to control image capture quality.
+* Add support for Torch and Auto Exposure
```
[the remark this answers is not in the document]
Comment (NeKoFu) [reply packages/camera/CHANGELOG.md]: I followed your recommendations and incremented the number

File: packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 4]
```diff
@@ -49,6 +49,8 @@
   private final Size captureSize;
   private final Size previewSize;
   private final boolean enableAudio;
+  private final boolean enableTorch;
```
[the remark this answers is not in the document]
Comment (NeKoFu) [reply packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java]: I followed your recommendations and replaced all "torch" by "flash".

File: packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 5]
```diff
@@ -49,6 +49,8 @@
...
+  private final boolean enableAE;
```
[the remark this answers is not in the document]
Comment (NeKoFu) [reply packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java]: Once again, I followed your recommendations and replaced abbreviations

File: packages/camera/lib/camera.dart (hunk 2 above)
Comment (NeKoFu) [reply packages/camera/lib/camera.dart]: I added more detailled comments.

File: packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java (hunk 1 above)
Comment (NeKoFu) [reply packages/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java]: This is how it works with "android.hardware.camera2" in the last Android SDK.
To change the state of the flash, we need to change the value of [FLASH_MODE](https://developer.android.com/reference/android/hardware/camera2/CaptureRequest.html#FLASH_MODE) through the [CaptureRequest](https://developer.android.com/reference/android/hardware/camera2/CaptureRequest.html).

**Comment by malikdoksoz** [comment]
When will it be added?

**Comment by malikdoksoz** [comment]
When release this update?

**Comment by mikerockett** [comment]
Hoping this will be released super-soon – our project is in _dire_ need of flash/torch access, which we can only achieve separately on iOS with the lamp package (Android doesn't allow multiple activities to use the camera at the same time, seemingly).

Thanks! 😃

**Comment by billowapp** [comment]
Really hoping this will be merged soon, it's the bread and butter of camera usage and will no doubt solve the need of many projects.

**Comment by mklim** [comment]
I'm taking this PR out of the backlog given the priority of the linked issue.

**Comment by mikerockett** [comment]
After playing around with this and looking at the implementation, it appears that the camera-request flags are not actually being set when the still image is being captured. As a result, turning on the flash is really just a torch function, which appears to be what this is PR for. This means that taking a still picture turns off the flash, seemingly because the mode (flash/ae/both - not actually sure as everyone seems to be doing different things) is not set correctly. Once the image is captured, the flash returns to is original state.

Guessing this applies to both platforms, though I'm unable to test until I'm back at work tomorrow. Have only tested on the Android side so far.

I've been reading up on a few things, and noticed the discussion of CameraX possibly coming into the Flutter world and I se a refactor is indeed underway. If I'm understanding this correctly, is the only limitation regarding being unable to use the flutter texture in the CameraX context?

With the project I'm working on, I'm not necessarily looking for all the advanced camera features, but simply looking to allow the user to set flash to on/off/always/auto.

**Comment by NeKoFu** [comment]
Thanks @mikerockett. I have found why the flash turns off when a picture is captured. It fixed now on Android and I will commit tomorrow after some change on the iOS side. My first goal was to use the flash as a torch but you right, it would be nice to control the flash for capture purpose only. So I added somes improvements.

**Comment by malikdoksoz** [comment]
We are waiting this update, very important for apps

**Comment by mikerockett** [comment]
@NeKoFu Thanks so much! I did attempt it myself, but couldn't get the right config going (between `FLASH_MODE` and `CONTROL_AE_MODE`, I just wasn't getting the right results). As an aside, I did look at other implementations, such as Google's Camera view and a few others. In doing so, I noticed that the Flutter camera plugin is built up somewhat differently to the others. For example, this one doesn't seem to keep state so that we can determine if the camera is ready to take a focused picture (something about needing pre-capture?). As a result, when I attempted to use flash on capture, I found that the images were over-exposed. That being said, I'm hopeful that CameraX will solve these issues due to it's high-level abstraction – perhaps we can get a more unified camera across all devices.

In the meantime, I've settled for allowing torch mode to carry over from preview to capture, after which the flash turns off. That said, if you update works better, I'll definitely switch to it :-)

**Comment by NeKoFu** [comment]
I share here with you my work on the Android version for feedbacks. This week, I didn't found enough time to work on the iOS part but I will syncs my code with the flutter:master branch and finish the iOS version asap.

**Comment by mikerockett** [comment]
Thanks @NeKoFu - much appreciated! As soon as I can, I'll move the updates into my fork and test it out.

**Comment by mikerockett** [comment]
I've done some initial testing, and it seems that `FlashMode.alwaysFlash` doesn't behave as it should. When instantiating a new `CameraController` with this mode, the flash turns on momentarily when the preview session starts up. Haven't dived into the code yet to see why this is happening, but a fair guess would be that you might be sharing all modes between preview and capture sessions. That's just a wild guess though.

When taking a picture in this mode, the file is saved before the camera has completed its capture lifecycle. In some lighting conditions, the initial flash occurs, at which point the image is saved, and then the camera continues to auto-focus and perform the final flash (I assume this is for ensuring the correct exposure?). As a result, the image is over-exposed. In other lighting conditions, the initial flash is not required, but the image is stored way before the capture is complete and so, as a result, the image does not benefit from flash at all.

Will post more as I go through the different modes. :)

**Comment by mikerockett** [comment]
`autoFlash` behaves somewhat differently, in that flash does not initiate when the preview boots, but only when the lighting conditions warrant it. If I initialise the camera whilst the camera is looking at dark lighting conditions (such as flat on a surface, or a dark room), the flash initiates on preview.

**Comment by malikdoksoz** [comment]
Why not update this feature?

**Comment by McCzarny** [comment]
If there are some difficulties with a flash, maybe You could consider splitting this task into 2 - one for an auto exposure and another for a flash? I'm sure that there are still a lot of people that needs just one of these features. (Including myself ;) )

**Comment by NeKoFu** [comment]
@mikerockett, thanks for your feedbacks. The flash turns on momentarily because the auto exposure mode is set to CONTROL_AE_MODE_ON_ALWAYS_FLASH. With this mode, auto exposure controls the flash and use it once immediately to adjust colors and light. Without this AE setting, I have no flash at all when I take a picture. It simply doesn't works on my phone without this AE mode and I have tested a lot of combinations. 😞 

I also saw one mistake with the continuous autofocus mode so I have fixed it.

For now, I finally have an old macbook so I can work on the iOS version more easily. 😋

**Comment by malikdoksoz** [comment]
> @mikerockett, thanks for your feedbacks. The flash turns on momentarily because the auto exposure mode is set to CONTROL_AE_MODE_ON_ALWAYS_FLASH. With this mode, auto exposure controls the flash and use it once immediately to adjust colors and light. Without this AE setting, I have no flash at all when I take a picture. It simply doesn't works on my phone without this AE mode and I have tested a lot of combinations. 😞
> 
> I also saw one mistake with the continuous autofocus mode so I have fixed it.
> 
> For now, I finally have an old macbook so I can work on the iOS version more easily. 😋

I tried iPhone 6-7-8( and plus) with iOS 12 and 13. It's working no problem :) I'm using this repo https://gitlab.com/mikerockett/flutter-camera . Can you give me recently updated repo (git link)

**Comment by mikerockett** [comment]
@NeKoFu – You’re most welcome. I do understand it can be quite difficult to get the right combination; I battled with this myself, which was extraordinarily frustrating given that I have very limited knowledge of how it all works. Thanks for working so hard on this.

@malikdoksoz – my repo is out of date, given the changes made here. I'm not going to be updating the repo at this point, as there is a commercial app in circulation that relies on it, and the flash behaviour used (always on / always off) is perfect for our needs at the moment.

**Comment by leonardodavila** [comment]
Any idea when this feature is gonna be merged into master?

**Comment by ghunkins** [comment]
Echoing @leonardodavila, this is a high priority feature for many ongoing initiatives. Can this feature be prioritized? It seems as though only small tweaks and a re-base are necessary.

**Comment by gisinator** [comment]
Cant wait for a working flash solution :)

**Comment by Ahmadre** [comment]
@NeKoFu  any updates?

File: packages/camera/example/lib/main.dart [hunk 6]
```diff
@@ -137,25 +138,123 @@ class _CameraExampleHomeState extends State<CameraExampleHome>
...
+          _toggleAudioWidget(),
+          _flashButton(),
+          _torchButton(),
         ],
       ),
     );
   }
 
+  /// Flash Toggle Button
+  Widget _flashButton() {
```
Comment (piotrek1543) [inline packages/camera/example/lib/main.dart]: I think that you can improve existing code by checking 'flashmode' using 'switch' statement. For now for the most of cases you're setting icon twice, which is not an expected behaviour.

**Comment by leonardodavila** [comment]
Is this code working for android and ios? Can i copy it?

**Comment by NeKoFu** [comment]
Hi everyone, I was very busy at the end of the year but I didn't forgot this package. At my company, I work now to porting our application to iOS and obviously, I need to finish the iOS part of this feature. 

I come back very soon with an update.

**Comment by malikdoksoz** [comment]
Can you give us some time?  Lack of flash is a big problem. Thank you

**Comment by felipebueno** [comment]
Hi, @NeKoFu!

Do you have any updates on this PR?

It's been 2 months since your last reply. We really need the flash feature ASAP :smile:

**Comment by GustavoFigueira** [comment]
Hi @NeKoFu!

Would be great to have this feature incorporated soon! My company needs tha flash function in our camera app.

Thanks in advance.

**Comment by orestesgaolin** [comment]
It seems like the title of the PR should include _Auto Focus_ not _Auto Exposure_

File: packages/camera/lib/camera.dart [hunk 7]
```diff
@@ -542,6 +616,24 @@ class CameraController extends ValueNotifier<CameraValue> {
...
+    if (!value.isInitialized || _isDisposed) {
+      throw CameraException(
+        'Uninitialized CameraController.',
+        'autoFocusMode was called on uninitialized CameraController',
+      );
+    }
+
+    try {
+      await _channel.invokeMethod<void>(
+          'setAutoFocus', <String, dynamic>{'mode': mode.index});
```
Comment (orestesgaolin) [inline packages/camera/lib/camera.dart:631]: This method is called `setAutoExposureMode:enableAutoExposure` in iOS project

**Comment by piotrek1543** [comment]
> It seems like the title of the PR should include _Auto Focus_ not _Auto Exposure_

The name is right. Check https://github.com/flutter/plugins/pull/1969/commits/1d8d34a0ed47701aee6637f5b25752b2510ac9cb

**Comment by malikdoksoz** [comment]
@NeKoFu When will I be able to use this?

**Comment by acoutts** [comment]
I just tried this out - it seems it is broken in iOS and a bit buggy:
- If you toggle the flash button it will eventually turn the torch on, but toggling the button again to `flash off` won't disable the torch unless you tap the torch button twice.
- When on the `auto` setting and the flash is turned on, it never turns off unless you again toggle the torch button twice.
- iOS is not working. Another issue is I see you are setting the torch setting on `AVCaptureDevice *device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];` and not using the actual active camera in use.
- Why is google sign on included in the PR?
- The dart code calls a method named `setFlash` on the platform channel but this isn't implemented on iOS, iOS only has explicit cases like `flashOn`, `flashOff` which is probably why it's broken in iOS.
- Throughout the code there are conflicting uses of the words `auto exposure` and `auto focus`, which are two very different functions of a camera.

I made a new PR before finding this one where I only implemented torch mode. It is working in iOS and Android and works with photos/video recording. I suggest anyone who needs the torch functionality now please try mine here: https://github.com/flutter/plugins/pull/2837

**Comment by mvanbeusekom** [comment]
closing as duplicate of #3338 and #3346

**Pull request closed without merging**